### ISSUE TITLE
Switch to proper logging instead of stdout

### ISSUE THREAD

**Issue by jbon** [OPEN-NEXT/wp2.2_dev#27]
use the python `logging` facility https://docs.python.org/3/library/logging.html
 - allows differenciating between info, warning, critical, debug
 - allows toggling outputs depending on wished information level
 - allows printing logs in different places at the same time (file or stdout)

**Comment by penyuan** [comment]
Yeah this has always been on my mind since seeing the old version at jbon/github-mining.

I will add provisions for this in my efforts to set up command line and configuration file parsing in issue #14.

**Comment by jbon** [comment]
I'm voluntary to implement this. I have already used it in the past so I am familiar with it. But I can understand it may create some merge conflicts with your development branch.

What do you think?

**Comment by jbon** [comment]
From de06e10 on in branch `iss27` loggers can be configured from the json configuration file. 

Documentation should be added in the `manual.md` introduced in issue #31 once this the branch `iss31` is merged back into master.

See here an example of logger configuration:

```
{
    "auth_token": ".token",
    "data_dir": "../__DATA__",
    "repo_list": "../data/tomine.csv",
    "create_data_dir": true,
    "log_config": {
        "version": 1,
        "disable_existing_loggers": false,
        "formatters": {
            "standard": {
                "format": "%(asctime)s - [%(levelname)s] %(name)s [%(module)s.%(funcName)s:%(lineno)d]: %(message)s", 
                "datefmt": "%Y-%m-%d %H:%M:%S"
            }
        },
        "handlers" : {
                "default": {
                    "class": "logging.StreamHandler",
                    "formatter": "standard"
                }
        }, 
        "loggers": {
                "__main__": { 
                    "handlers" : ["default"], 
                    "propagate": false
                }
        },
        "root": {
                "level": "INFO",
                "handlers": ["default"]
        }
    }
}
```

Mining the repo `OPEN-NEXT/wp2.2_dev`, this yields the following log:
```
INFO:root:Trying to parse configuration file:config.json
INFO:root:GitHub authentication key looks OK.
2020-07-03 19:36:14 - [INFO] root [get_Github_forks.get_Github_forks:44]: Finding forks of OPEN-NEXT/wp2.2_dev page 1
2020-07-03 19:36:14 - [INFO] root [start.main:85]: There are 1 forks of OPEN-NEXT/wp2.2_dev
2020-07-03 19:36:14 - [INFO] root [start.main:95]: retrieving commits in OPEN-NEXT/wp2.2_dev
2020-07-03 19:36:14 - [INFO] root [get_commits.get_commits:44]: fetching info at https://github.com/OPEN-NEXT/wp2.2_dev
2020-07-03 19:36:14 - [INFO] perceval.backends.core.git [git.__fetch_commits_from_repo:285]: Fetching commits: 'https://github.com/OPEN-NEXT/wp2.2_dev' git repository from 1970-01-01 00:00:00+00:00 to 2100-01-01 00:00:00+00:00; all branches
2020-07-03 19:36:16 - [INFO] perceval.backends.core.git [git.fetch_items:163]: Fetch process completed: 136 commits fetched
```

**Comment by jbon** [comment]
to avoid getting spammed from logs of third party modules (e.g. perceval), add the following to the config
```
        "loggers": {
                "__main__": { 
                    "handlers" : ["default"], 
                    "propagate": false
                },
                "perceval.backend": {
                    "level": "INFO"         
                },
                "perceval.backends.core.git": {
                    "level": "INFO"         
                },
                "urllib3.connectionpool": {
                    "level": "INFO"         
                }                              
        },
```

The config file needs to be compliant with the [logging config dict schema](https://docs.python.org/3/library/logging.config.html#logging-config-dictschema)

**Comment by jbon** [comment]
As far as I can judge, the issue can be closed from a49c7bb on. I rearranged the logger calls throughout the modules to yield the following logs:

```
INFO:root:Trying to parse configuration file:config.json
2020-07-03 20:38:41 - [DEBUG] root [initialise.initialise_options:143]: GitHub authentication key looks OK.
2020-07-03 20:38:41 - [INFO] root [start.main:65]: Start processing the 3 repositories passed on
2020-07-03 20:38:41 - [INFO] root [start.main:67]: --- Start processing repository OPEN-NEXT/WP2.2_dev ---
2020-07-03 20:38:41 - [INFO] root [start.main:86]: 0 forks found
2020-07-03 20:38:42 - [INFO] root [start.main:104]: 137 commits found
2020-07-03 20:38:43 - [INFO] root [start.main:135]: Commit history built with 137 nodes and 154 edges
2020-07-03 20:38:43 - [INFO] root [start.main:153]: File change history built with 32 files and 211 file changes
2020-07-03 20:38:43 - [INFO] root [start.main:170]: Commiter graph built with 6 unique committers
2020-07-03 20:38:43 - [INFO] root [start.main:67]: --- Start processing repository OPEN-NEXT/WP2.2_reference ---
2020-07-03 20:38:45 - [INFO] root [start.main:86]: 2 forks found
2020-07-03 20:38:47 - [INFO] root [start.main:104]: 29 commits found
2020-07-03 20:38:47 - [INFO] root [start.main:135]: Commit history built with 29 nodes and 33 edges
2020-07-03 20:38:47 - [INFO] root [start.main:153]: File change history built with 5 files and 32 file changes
2020-07-03 20:38:47 - [INFO] root [start.main:170]: Commiter graph built with 4 unique committers
2020-07-03 20:38:47 - [INFO] root [start.main:67]: --- Start processing repository jbon/github-mining ---
2020-07-03 20:38:48 - [INFO] root [start.main:86]: 3 forks found
2020-07-03 20:38:51 - [INFO] root [start.main:104]: 54 commits found
2020-07-03 20:38:51 - [INFO] root [start.main:135]: Commit history built with 54 nodes and 59 edges
2020-07-03 20:38:51 - [INFO] root [start.main:153]: File change history built with 21 files and 74 file changes
2020-07-03 20:38:51 - [INFO] root [start.main:170]: Commiter graph built with 4 unique committers
```

with the following config file
```
{
    "auth_token": ".token",
    "data_dir": "../__DATA__",
    "repo_list": "../data/tomine.csv",
    "create_data_dir": true,
    "log_config": {
        "version": 1,
        "disable_existing_loggers": false,
        "formatters": {
            "standard": {
                "format": "%(asctime)s - [%(levelname)s] %(name)s [%(module)s.%(funcName)s:%(lineno)d]: %(message)s", 
                "datefmt": "%Y-%m-%d %H:%M:%S"
            }
        },
        "handlers" : {
                "default": {
                    "class": "logging.StreamHandler",
                    "formatter": "standard"
                }
        }, 
        "loggers": {
                "__main__": { 
                    "handlers" : ["default"], 
                    "propagate": false
                },
                "perceval.backend": {
                    "level": "CRITICAL"         
                },
                "perceval.backends.core.git": {
                    "level": "CRITICAL"         
                },
                "urllib3.connectionpool": {
                    "level": "CRITICAL"         
                }                                
        },
        "root": {
                "level": "DEBUG",
                "handlers": ["default"]
        }
    }
}
```

and repo_list
```
owner,repo
OPEN-NEXT,WP2.2_dev
OPEN-NEXT,WP2.2_reference
jbon,github-mining
```

@penyuan  for review

**Comment by penyuan** [comment]
Sorry I didn't realise that I didn't merge it, but it has been merged now. Closing this issue for now. Thanks!
